api/server/middleware.go

```diff
@@ -0,0 +1,99 @@
+// Any function that has the appropriate signature can be register as a middleware.
+type middleware func(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error)
+
+// loggingMiddleware logs each request when logging is enabled.
+func (s *Server) loggingMiddleware(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error) {
+	if s.cfg.Logging {
+		logrus.Infof("%s %s", r.Method, r.RequestURI)
+	}
+	return ctx, nil
+}
+
+// requestIDMiddleware generates a uniq ID for each request.
+// This ID travels inside the context for tracing purposes.
+func requestIDMiddleware(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error) {
+	reqID := stringid.TruncateID(stringid.GenerateNonCryptoID())
+	return context.WithValue(ctx, context.RequestID, reqID), nil
+}
+
+// userAgentMiddleware checks the User-Agent header looking for a valid docker client spec.
+func (s *Server) userAgentMiddleware(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error) {
+	if strings.Contains(r.Header.Get("User-Agent"), "Docker-Client/") {
+		dockerVersion := version.Version(s.cfg.Version)
+
+		userAgent := strings.Split(r.Header.Get("User-Agent"), "/")
+
+		// v1.20 onwards includes the GOOS of the client after the version
+		// such as Docker/1.7.0 (linux)
+		if len(userAgent) == 2 && strings.Contains(userAgent[1], " ") {
+			userAgent[1] = strings.Split(userAgent[1], " ")[0]
+		}
+
+		if len(userAgent) == 2 && !dockerVersion.Equal(version.Version(userAgent[1])) {
+			logrus.Debugf("Warning: client and server don't have the same version (client: %s, server: %s)", userAgent[1], dockerVersion)
+		}
+	}
+	return ctx, nil
+}
+
+// corsMiddleware sets the CORS header expectations in the server.
+func (s *Server) corsMiddleware(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error) {
+	// If "api-cors-header" is not given, but "api-enable-cors" is true, we set cors to "*"
+	// otherwise, all head values will be passed to HTTP handler
+	corsHeaders := s.cfg.CorsHeaders
+	if corsHeaders == "" && s.cfg.EnableCors {
+		corsHeaders = "*"
+	}
+
+	if corsHeaders != "" {
+		writeCorsHeaders(w, r, corsHeaders)
+	}
+	return ctx, nil
+}
+
+// versionMiddleware checks the api version requirements before passing the request to the server handler.
+func versionMiddleware(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) (context.Context, error) {
+	apiVersion := version.Version(vars["version"])
+	if apiVersion == "" {
+		apiVersion = api.Version
+	}
+
+	if apiVersion.GreaterThan(api.Version) {
+		return ctx, errors.ErrorVersionMismatch.WithArgs(fmt.Sprintf("client is newer than server (client API version: %s, server API version: %s)", apiVersion, api.Version))
```

Comment:
move these strings into unique error objects.  When done we shouldn't have any fmt.Errorf() or Sprintf() in error generation code.  So I think this would become:
```

return ctx, errors.ErrorClientIsNewer.WithArgs(apiVersion, api.Version)
```
where ErrorClientIsNewer has the Message value of:
```
"client is newer than server (client API version: %s, server API version: %s)"
```